### DISCUSSION TITLE
Why is GIMP not using persist, cannot be started via shim

### DISCUSSION THREAD

**Discussion by ltguillaume** [ScoopInstaller/Extras#12271] [Q&A]
1. Why is GIMP not storing the contents of `%AppData%\GIMP` into `scoop\persist\gimp`?
2. Additionally, the updates check could be disabled by writing `(check-updates no)` to the file `scoop\persist\gimp\2.10\gimprc`.
3. Why do I only get a console window with the following errors if I try to start GIMP via the `gimp.exe` shim?
```
GEGL-Message: 19:56:52.116: Module 'D:\Scoop\apps\gimp\current\lib\gegl-0.4\exr-save.dll' load error: 'D:\Scoop\apps\gimp\current\lib\gegl-0.4\exr-save.dll': The specified module could not be found.
GEGL-Message: 19:56:52.117: Module 'D:\Scoop\apps\gimp\current\lib\gegl-0.4\exr-load.dll' load error: 'D:\Scoop\apps\gimp\current\lib\gegl-0.4\exr-load.dll': The specified module could not be found.
GIMP-Error: Unable to run plug-in "webexport"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\webexport)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "wavelet-denoise"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\wavelet-denoise)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "streak"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\streak)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "separate_import"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\separate_import)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "separate"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\separate)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "resynthesizer-gui"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\resynthesizer-gui)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "resynthesizer"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\resynthesizer)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "plug_in_lqr_iter"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\plug_in_lqr_iter)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "gimpmask"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\gimpmask)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "gimp-lqr-plugin"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\gimp-lqr-plugin)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "fix-ca"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\fix-ca)

Failed to execute child process (No such file or directory)

GIMP-Error: Unable to run plug-in "dbp"
(C:\Users\MyAccount\AppData\Roaming\GIMP\2.10\plug-ins\dbp)

Failed to execute child process (No such file or directory)

gimp_check_updates_callback: loading of https://www.gimp.org/gimp_versions.json failed: GET request failed: WinHttp error: CANNOT_CONNECT
```

**Comment by goyalyashpal** [comment]
> 1. Why is GIMP not storing the contents of `%AppData%\GIMP` into `scoop\persist\gimp`?

in my naive understanding of gimp; the choice to use `%...%` as default location for configs is on the app developers. as far as i know, scoop only manages the persists which have some chance to work from the same directory of that app's install

**Reply by ltguillaume** [reply]
So GIMP isn't portable and should be `gimp-np`, unless "non-portable" has its own meaning in the context of Scoop (i.e. that Scoop can't contain the _program files_ inside its own directory and doesn't care about the _user data_ in this). I doubt this is the case.

Since Scoop is already making extensive use of directory junctions, it could do the same for `%(Local)AppData%` folders. A few days ago I wrote a manifest doing just this for Dino: https://github.com/ltguillaume/schep/blob/main/bucket/dino.json

- So, on install/update, if `%LocalAppData%\Dino` doesn't exist or is empty, it creates a junction there to Dino's `persist_dir`.
- On uninstall, it checks if `%LocalAppData%\Dino` is a junction to `persist_dir` and removes the junction if so.

**Reply by goyalyashpal** [reply]
> unless "non-portable" has its own meaning in the context of Scoop

oh lol. but yes and no. 

the portable generally means that usable via archives. and it means same in case of scoop too. 
the meaning of portable u are using doesn't have a dedicated name afaik.

consequently, the non-portable bucket is, based on observation of apps it offers, for the apps which do not offer even an archive for their installation. that is, their installation is solely limited to and managed by the executables their authors provided them with.


example:

https://github.com/ScoopInstaller/Nonportable/blob/master/bucket/autohotkey-np.json#L6

```json
"url": "https://github.com/Lexikos/AutoHotkey_L/releases/download/v2.0.12/AutoHotkey_2.0.12_setup.exe#/setup.exe",
```

**Reply by ltguillaume** [reply]
Really? Well, for starters, AutoHotkey is a bad example: it offers zip-files next to the installers (https://github.com/AutoHotkey/AutoHotkey/releases/download/v2.0.12/AutoHotkey_2.0.12.zip) *and* it allows the parameters `/to` or `/installto` to specify the installation folder when using th installer.

Second, creating your own definition of a widely used term is very confusing.

**Reply by goyalyashpal** [reply]
> Well, for starters, AutoHotkey is a bad example: it offers zip-files next to the installers

consequently, the zip offering of the ahk is available in this bucket too.

```
$ scoop-search autohotkey
'extras' bucket:
    autohotkey (2.0.12)

'nonportable' bucket:
    autohotkey-np (2.0.12)
```

> Second, creating your own definition of a widely used term is very confusing.

nobody created new definitions of "widely used" terms. i told what i saw out in the wild over all these years.

> it allows the parameters `/to` or `/installto` to specify the installation folder when using th installer.

that doesn't mean the installer is portable. it just means the installer is `cli` friendly. that installer is still an `executable` file.

**Reply by ltguillaume** [reply]
>nobody created new definitions of "widely used" terms. i told what i saw out in the wild over all these years.

Well, the ambiguity that Scoop adds to the term "portable" is quite significant imho.

> that doesn't mean the installer is portable. it just means the installer is cli friendly. that installer is still an executable file.

Well in that case, you're using the normal definition of "portable" (i.e. that modifications are being made to the system itself), not the definition we just considered to accept as being used (solely) by Scoop _(i.e. that Scoop can't contain the program files inside its own directory and doesn't care about the user data in this)_. If Scoop is actually using both definitions, then (back to the initial issue) either GIMP should be `gimp-np` or there shouldn't be a distinction made between `autohotkey` and `autohotkey-np`, even if there in fact were a difference (e.g. that only `autohotkey-np` would add file associations).

**Reply by goyalyashpal** [reply]
> Well, the ambiguity that Scoop adds to the term "portable" is quite significant imho.

to me, it aligns with what i saw out in the wild; and i didn't even have to make a second guess to it. it was what i interpretted, and on checking it out, it aligned well. and i myself have just been using scoop for as recent as 2 weeks - so, there's no "used to it bias" here.

in fact, if there is an `<app>.zip` used in a manifest inside the `np` bucket, that would be ambiguous and surprise to me.

**Reply by goyalyashpal** [reply]
> definition ... being used (solely) by Scoop _(i.e. that Scoop can't contain the program files inside its own directory and doesn't care about the user data in this)_.

can u rephrase a bit? i didn't quite follow.

**Reply by ltguillaume** [reply]
Then I invite you to read the first post of https://github.com/ScoopInstaller/Scoop/discussions/5389

**Reply by goyalyashpal** [reply]
> first post of [ScoopInstaller/Scoop#5389](https://github.com/ScoopInstaller/Scoop/discussions/5389)

![image](https://github.com/ScoopInstaller/Extras/assets/19423063/379c40d8-1ff9-4970-b018-44a5bd99dbc0)

can u link directly to the comment/post u want me to read? there are 4 possibilities of the first here (description x1 + sorted by x3)

**Reply by ltguillaume** [reply]
>can u rephrase a bit? i didn't quite follow.

In most cases, Scoop doesn't care about the program's user data: where it resides, what happens to it when a program is uninstalled via Scoop. Those issues are, however, an obvious and inherent part of the term "portable" (think of "data portability").

Non-portable in the eyes of Scoop seems to be either of these cases only:
- The installation requires admin permissions
- The installation does not allow for (or the manifest hasn't been made to put the) installation into the `scoop\apps` folder

That is a significant difference.

> can u link directly to the comment/post u want me to read? there are 4 possibilities of the first here (description x1 + sorted by x3)

Not sure what you mean, but you've already rolleyed it: https://github.com/ScoopInstaller/Scoop/discussions/5389#discussion-4863970

**Reply by goyalyashpal** [reply]
> rolleyed

oh lol, that was not "roll eye". roll eye is this: `🙄`
i use the eyes i.e. `👀`  to indicate, not to others, but to myself for later, that huh, this is interesting, and worth "looking" again too

**Comment by goyalyashpal** [comment]
> 3\. Why do I only get a console window with the following errors if I try to start GIMP via the `gimp.exe` shim?

i just tried doing that, and i also get a console window. but it was blank - no errors. and gimp didn't launch.

but i am curious, why would u use the bin-shim `gimp.exe` and not the shortcut `GIMP` ?

**Reply by ltguillaume** [reply]
1. I was checking out all features of Scoop
2. If I'm using a console window at the time I want to start GIMP
3. When I'd like to directly open a file in the current folder of a console in GIMP
4. I usually add two-letter shortcuts to programs which I run using a custom "Run..." dialog (by adding `HKLM\Software\Microsoft\Windows\CurrentVersion\App Paths\%keyword%.exe` keys) and I tried that using the shim

**Reply by goyalyashpal** [reply]
> 3\. When I'd like to directly open a file in the current folder of a console in GIMP

~does gimp even allow that? many GUI applications lack a CLI, or have a bad one.~

trying tab completion on the shell, i see there's a `gimp-console.exe`. maybe u want to try out with that one.

_update:_ yeah, `gimp-console --help` confirms that. find the abridged output of it attached alongwith.
u might be interested in using `gimp-console.exe --no-splash`

_update2:_ but `gimp --help` also shared the same message (and also said `gimp-console` in usage... i.e. it is probably pointing to `gimp-console` when running from cli). 
and running `gimp.exe -s` or even `gimp-console -s` or `-a` from the win 10 start menu or even from a terminal+shell... the results were same: blank process with no output, no launches.

```console
$ gimp<tab><tab>
gimp-console-2.10.exe  gimp-console.exe       gimp.exe               gimptool-2.0.exe       gimptool.exe

$ gimp-console --help
Usage:
  gimp-console-2.10.exe [OPTION?] [FILE|URI...]

GNU Image Manipulation Program

Help Options:
  -h, --help                              Show help options
  --help-all                              Show all help options
  --help-gegl                             Show GEGL Options

Application Options:
  -v, --version                           Show version information and exit
  --license                               Show license information and exit
  --verbose                               Be more verbose
  -n, --new-instance                      Start a new GIMP instance
  -a, --as-new                            Open images as new
  -i, --no-interface                      Run without a user interface
  -d, --no-data                           Do not load brushes, gradients, patterns, ...
  -f, --no-fonts                          Do not load any fonts
  -s, --no-splash                         Do not show a splash screen
  ...
```

**Reply by ltguillaume** [reply]
The `gimp.exe` shim already starts `gimp-console.exe` (see `gimp.shim`).

As such, there's no difference.
